Replaced diagram to latest one and added description (#53)
klaytn/kips: KIPs/kip-34.md

This is the overview of the common architecture of Klaytn SDK.

> ![whole_20210217](https://user-images.githubusercontent.com/32922423/108157902-2a2a2380-7127-11eb-814f-b166f2983e9f.png)

Reviewer: For both images (this and the one below), in the KIP-17 box, next to "MintwithTokens" there is a little typo "toeken"

Author: I fixed mistypo :)

`Keystore` is a class that contains encrypted keyring. The internally defined value differs depending on the keystore version, see [KIP-3](./kip-3.md).

> In the wallet layer, an interface called `IWallet` is defined, and this interface is an interface of `Wallet` that stores data used when signing a transaction.

Reviewer 2: ```suggestion
`IWallet` is an interface that contains common functions related to signing a transaction.
```

Is there the class `Wallet` somewhere?

Author: There is not separated `Wallet` class. I used the term Wallet as the concept of a class that stores data used when signing. In caver, Wallet corresponds to KeyringContainer. 

I described here with an `account`, which is a data unit stored here, and a `wallet` that manages and stores it, but if there is a more clear description method, please tell me. ;-;

Reviewer 2: If wallet is not an actual class, it would be better to use wallet instead of `wallet`.
I think my suggestion is enough to explain `IWallet`. What do you think?

Author: Oh, I understood what you were saying. It looked like a real class by processing a block of code. Including the suggestions given above, I uploaded a separate [commit](https://github.com/klaytn/kips/pull/53/commits/f3152a29e248009a39314722a33b1e960d400f12) and modified it.

| Method | Description |
| ----------- | ----------- |

> | getPublicKey(): List&#60;String&#62; | Returns the public key strings. |

Reviewer 2: @jimni1222 Is this function remained?

Author: This is remained because `compressed` parameter is optional. 
For java, to support optional parameter, function should be overloaded.

> | getPublicKey(compressed: Boolean): String | Returns the compressed public key strings if compressed is true. It returns the uncompressed public key strings otherwise. |

Reviewer 2: ```suggestion
| getPublicKey(compressed: Boolean): List&#60;String&#62; | Returns the compressed public key strings if compressed is true. It returns the uncompressed public key strings otherwise. |
```

> | getKeyByRole(role: int): List&#60;PrivateKey&#62; | Returns the keys specified by the role. |

Reviewer 2: ```suggestion
| getKeyByRole(role: int): List&#60;PrivateKey&#62; | Returns the keys specified by the role. MultipleKeyring returns the same keys since it does not have role. |
```

| Method | Description |
| ----------- | ----------- |
| getPublicKey(): List&#60;List&#60;String&#62;&#62; | Returns the public key strings for all roles. |

> | getPublicKey(compressed: Boolean): String | Returns the compressed public key strings for all roles if compressed is true. It returns the uncompressed public key strings otherwise. |

Reviewer 2: ```suggestion
| getPublicKey(compressed: Boolean): List&#60;List&#60;String&#62;&#62; | Returns the compressed public key strings for all roles if compressed is true. It returns the uncompressed public key strings otherwise. |
```

> | getKeyByRole(role: int): List&#60;PrivateKey&#62; | Returns the private keys for all roles. |

Reviewer 2: ```suggestion
| getKeyByRole(role: int): List&#60;PrivateKey&#62; | Returns the private keys of the given role. |
```

| createWithRoleBasedKey(address: String, roleBasedKeys: List&#60;String[]&#62;): RoleBasedKeyring | Creates a RoleBasedKeyring instance from an address and a 2D array of which each array element contains keys defined for each role. It throws an exception if the address string or the private key strings are invalid. |
| decrypt(keystore: Object, password: String): Keyring | Decrypts a keystore v3 or v4 JSON and returns the decrypted keyring instance. It throws an exception if the decryption is failed. |

#### IWallet

> `IWallet` is the interface of Wallet that manages data used when signing transactions. All Wallet classes must implement `IWallet`.

Reviewer 2: ```suggestion
`IWallet` is an interface that contains common functions related to signing a transaction.. All Wallet classes must implement `IWallet`.
```

> `IWallet` is an interface that contains common functions related to signing a transaction.. All Wallet classes must implement `IWallet`.

Reviewer 3: ```suggestion
`IWallet` is an interface that contains common functions related to signing a transaction. All Wallet classes must implement `IWallet`.
```

> Caver has a `KeyringContainer` instance that implements `IWallet`, and if it is a Wallet class that implements `IWallet` interface, it can be used in Caver instead of `KeyringContainer`.

Reviewer 2: ```suggestion
Caver has a class `KeyringContainer` that implements `IWallet`. If it is a class that implements `IWallet` interface, it can be used instead of `KeyringContainer` to sign a transaction.
```

##### Variables

None

##### Methods

| Method | Description |
| ----------- | ----------- |
| generate(num: int): List<String> | Generates `num` SingleKeyring instances and stores them in the `Wallet` instance. |

> | sign(address: String, transaction: AbstractTransaction): AbstractTransaction | Signs the transaction as a sender of the transaction and appends signatures in the transaction object using the account associated with the given address in the `Wallet` instance. It throws an exception if the account to be used for signing cannot be found in the `Wallet` instance or if `address` is different from the sender address of `transaction`. |

Reviewer 2: ```suggestion
| sign(address: String, transaction: AbstractTransaction): AbstractTransaction | Signs the `transaction` using the private key of the `address` and appends signatures in the `transaction` object. It throws an exception if the `address` is not found or the `address` and the sender address of `transaction` are not the same. |
```

Author: I applied in https://github.com/klaytn/kips/pull/53/commits/6b1c2acde7f0aa1970bfa8aecd0e018cb3f31504

> | signAsFeePayer(address: String, transaction: AbstractFeeDelegatedTransaction): AbstractFeeDelegatedTransaction | Signs the transaction as a fee payer of the transaction and appends signatures in the transaction object using the account associated with the given address in the `Wallet` instance. It throws an exception if the account to be used for signing cannot be found in the `Wallet` instance or if the address received as a parameter and the fee payer of the actual transaction do not match. |

Reviewer 2: ```suggestion
| signAsFeePayer(address: String, transaction: AbstractFeeDelegatedTransaction): AbstractFeeDelegatedTransaction | Signs the `transaction` using the private key of the `address` as a fee payer and appends signatures in the transaction object. It throws an exception if the `address` is not found or the `address` and the fee payer of the `transaction` are not the same. |
```

Author: I applied in https://github.com/klaytn/kips/pull/53/commits/6b1c2acde7f0aa1970bfa8aecd0e018cb3f31504

| isExisted(address: String): Boolean | Returns whether the account corresponding to the address exists. |

> | remove(address: String): Boolean | Deletes an account associated with the given address from the `Wallet` instance. |

Reviewer 2: ```suggestion
| remove(address: String): Boolean | Deletes an account associated with the given address. |
```

Author: I applied in here https://github.com/klaytn/kips/pull/53/commits/6b1c2acde7f0aa1970bfa8aecd0e018cb3f31504

#### KeyringContainer

> `KeyringContainer` is a class that contains SingleKeyring, MultipleKeyring, and RoleBasedKeyring instances based on the address. `KeyringContainer` must implement the `IWallet` interface, and manages SingleKeyring, MultipleKeyring, and RoleBasedKeyring instances inside.

Reviewer 2: ```suggestion
`KeyringContainer` is a class that can contain SingleKeyring, MultipleKeyring, and RoleBasedKeyring instances based on the address. `KeyringContainer` implements `IWallet`.
```

| newKeyring(address: String, roleBasedPrivateKeyArray: List&#60;List&#60;String&#62;&#62;): AbstractKeyring | Creates a `RoleBasedKeyring` instance with given parameters and adds it to the `KeyringContainer` instance. It returns the newly added keyring instance. It throws an exception if the address string or private key strings are invalid. |
| updateKeyring(keyring: AbstractKeyring): AbstractKeyring | Replaces the keyring instance having the same address with the given keyring in the parameter. It throws an exception if the matching keyring is not found. |
| getKeyring(address: String): AbstractKeyring | Returns the keyring instance corresponding to the address. |

> | sign(address: String, transaction: AbstractTransaction): AbstractTransaction | Signs the transaction as a sender of the transaction and appends signatures in the transaction object using the keyring associated with the given address in the `KeyringContainer` instance. This method will use all the private keys in the keyring. It throws an exception if the keyring to be used for signing cannot be found in the `KeyringContainer` instance or if `address` is different from the sender address of `transaction`. |

Reviewer 2: Why is this function removed?

Author: I have excluded the functions defined in IWallet above. However, since it is correct to be included, I added it again in [this commit](https://github.com/klaytn/kips/pull/53/commits/6b1c2acde7f0aa1970bfa8aecd0e018cb3f31504).

The `KIP17` class provides the functions to interact with [KIP-17] token contracts on Klaytn. This class allows users to easily
deploy and execute [KIP-17] token contracts on Klaytn. `KIP17` maps all functions defined in [KIP-17] and provides them as class methods.

> In the future, for [KCT](http://kips.klaytn.com/token) that are additionally defined in [KIP](http://kips.klaytn.com), APIs that support the newly added KCT will be implemented in Caver. For the KCT added to KIP, please refer to [here](http://kips.klaytn.com/token).

Reviewer 2: ```suggestion
More token standards defined in [KCT](http://kips.klaytn.com/token) can be implemented here. For other KCT implementations can be found in the SDK references ([caver-js](https://docs.klaytn.com/bapp/sdk/caver-js/api-references/caver.kct), [caver-java](https://javadoc.io/doc/com.klaytn.caver/core/latest/index.html)).
```
